### PR TITLE
Addon-docs/Angular: Render user defined template as source if it exists

### PR DESCRIPTION
**Pull request by yngvebn**
Issue: https://github.com/storybookjs/storybook/issues/14268

## What I did

Added a  flag to indicate that the story has a custom template, and render that instead of a generated template

## How to test

- Is this testable with Jest or Chromatic screenshots?
- Does this need a new example in the kitchen sink apps?
- Does this need an update to the documentation?

If your answer is yes to any of these, please make sure to include it in your PR.

Added a new story for the doc-button component, which uses a template. Also several other stories have custom templates, and can be used for verification

### CODE REVIEW

File: app/angular/src/client/preview/decorateStory.ts [hunk 1]
```diff
@@ -43,13 +42,15 @@ const prepareMain = (
...
+  const noTemplate = hasNoTemplate(template);
 
-  if (hasNoTemplate(template) && component) {
+  if (noTemplate && component) {
     template = computesTemplateFromComponent(component, story.props, '');
   }
   return {
     ...story,
     ...(template ? { template } : {}),
+    hasCustomTemplate: !noTemplate,
```
Comment (ThibaudAV) [inline app/angular/src/client/preview/decorateStory.ts:53]:
```suggestion
    ...(template ? { template, hasCustomTemplate: !noTemplate } : {}),
```
no ?

File: app/angular/src/client/preview/types.ts [hunk 2]
```diff
@@ -28,4 +28,5 @@ export interface StoryFnAngularReturnType {
   moduleMetadata?: NgModuleMetadata;
   template?: string;
   styles?: string[];
+  hasCustomTemplate?: boolean;
```
Comment (ThibaudAV) [inline app/angular/src/client/preview/types.ts:31]: Isn't there a way to set this flag in the story context? what do you think @shilman ?

**Comment by ThibaudAV** [comment]
If the code is ok for everyone
I'll look to fix the tests and add a new one to avoid future regression

**Comment by yngvebn** [comment]
> If the code is ok for everyone
> 
> I'll look to fix the tests and add a new one to avoid future regression 

I'll fix the code like you proposed first 👍

File: app/angular/src/client/preview/types.ts (hunk 2 above)
Comment (yngvebn) [reply app/angular/src/client/preview/types.ts]: Any similar stuff being done? I could have a look, if I know where to :)

File: app/angular/src/client/preview/types.ts [hunk 3]
```diff
@@ -28,5 +28,5 @@ export interface StoryFnAngularReturnType {
   moduleMetadata?: NgModuleMetadata;
   template?: string;
   styles?: string[];
-  hasCustomTemplate?: boolean;
+  userDefinedTemplate?: boolean;
```
Comment (shilman) [inline app/angular/src/client/preview/types.ts:31]: can we do without adding this, and instead just look at the value of `template`?

Comment (yngvebn) [reply app/angular/src/client/preview/types.ts:31]: That's the problem, if template is missing, it's autogenerated. That's why I have to set this flag prior to the autogeneration...

Comment (yngvebn) [reply app/angular/src/client/preview/types.ts:31]: @shilman The source-decorator happens so late in the flow, and a template is required for rendering the component. I guess the template to be rendered could possibly be a more explicit property, leaving `template` as the actual template provided in the story. Don't really see how that would improve this particular code though.

Comment (shilman) [reply app/angular/src/client/preview/types.ts:31]: Gotcha, thanks for clarifying!

**Comment by yngvebn** [comment]
@ThibaudAV / @shilman Seems like I finally managed to get the unit-tests to play along. This should be good for another review

**Review by shilman — APPROVED** [review]
LGTM! 🚀

**Pull request merged**